src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/MemberMatch/MemberMatchService.cs

```diff
@@ -0,0 +1,174 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Extensions.DependencyInjection;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Core.Features.Definition;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions.Parsers;
+using Microsoft.Health.Fhir.Core.Models;
+using Expression = Microsoft.Health.Fhir.Core.Features.Search.Expressions.Expression;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.MemberMatch
+{
+    public sealed class MemberMatchService : IMemberMatchService
+    {
+        private readonly Func<IScoped<ISearchService>> _searchServiceFactory;
+        private readonly IResourceDeserializer _resourceDeserializer;
+        private readonly ISearchIndexer _searchIndexer;
+        private readonly IExpressionParser _expressionParser;
+        private readonly SearchParameterInfo _coverageBeneficiaryParameter;
+        private readonly ILogger<MemberMatchService> _logger;
+
+        public MemberMatchService(
+            Func<IScoped<ISearchService>> searchServiceFactory,
+            IResourceDeserializer resourceDeserializer,
+            ISearchIndexer searchIndexer,
+            ISearchParameterDefinitionManager.SearchableSearchParameterDefinitionManagerResolver searchParameterDefinitionManagerResolver,
+            IExpressionParser expressionParser,
+            ILogger<MemberMatchService> logger)
+        {
+            EnsureArg.IsNotNull(searchServiceFactory, nameof(searchServiceFactory));
+            EnsureArg.IsNotNull(resourceDeserializer, nameof(resourceDeserializer));
+            EnsureArg.IsNotNull(searchIndexer, nameof(searchIndexer));
+            EnsureArg.IsNotNull(searchParameterDefinitionManagerResolver, nameof(searchParameterDefinitionManagerResolver));
+            EnsureArg.IsNotNull(expressionParser, nameof(expressionParser));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _searchServiceFactory = searchServiceFactory;
+            _resourceDeserializer = resourceDeserializer;
+            _searchIndexer = searchIndexer;
+            _expressionParser = expressionParser;
+            _coverageBeneficiaryParameter = searchParameterDefinitionManagerResolver().GetSearchParameter("Coverage", "beneficiary");
```

Comment:
I don't think this should be part of the constructor since there is a chance `GetSearchParameter` can throw an exception (if the resource type is not found). Another option is to use `TryGetSearchParameter` maybe?

Reply:
if it's not present we will throw exception and member-match service would return 
```
 "resourceType": "OperationOutcome",
    "id": "c5821932-add9-4de9-ae42-2089a383b04b",
    "issue": [
        {
            "severity": "error",
            "code": "not-supported",
            "diagnostics": "The search parameter 'beneficiary' is not supported for resource type 'Coverage'."
        }
    ]
```

which I think is ok...